doc/source/guide/derivatives_guide.rst

```diff
@@ -0,0 +1,155 @@
+
+.. math::
+    f'(x)(d) = \lim_{h \rightarrow 0} \frac{f(x + dh) - f(x)}{h}.
+
+Here we see (as implied by the notation) that :math:`f'(x)` is actually an operator
+
+.. math::
+    f'(x) : \mathbb{R}^n \rightarrow \mathbb{R}^m.
+
+We can rewrite this using the explicit requirement that :math:`f'(x)` is a linear approximation of :math:`f` at :math:`x`, i.e.
+
+.. math::
+   \lim_{\| d \| \rightarrow 0} \frac{\| f(x + d) - f(x) - f'(x)(d) \|}{\| d \|} = 0
+
+This notion naturally extends to an `Operator` :math:`f : \mathcal{X} \rightarrow \mathcal{Y}` between Banach spaces :math:`\mathcal{X}` and :math:`\mathcal{Y}` with norms :math:`\| \cdot \|_\mathcal{X}` and :math:`\| \cdot \|_\mathcal{Y}`, respectively.
+Here :math:`f'(x)` is defined as the linear operator (if it exists) that satisfies
+
+.. math::
+   \lim_{\| d \| \rightarrow 0} \frac{\| f(x + d) - f(x) - f'(x)(d) \|_\mathcal{Y}}{\| d \|_\mathcal{X}} = 0
+
+This definition of the derivative is called the *Fréchet derivative*.
+
+Rules for the Fréchet derivative
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Many of the usual rules for derivatives also hold for the Fréchet derivative, i.e.
+
+* Linearity
+
+.. math::
+    (a f + b y)'(x)(y) = a f'(x)(y) + b g'(x)(y)
+
+* Chain rule
+
+.. math::
+    (g \circ f)'(x)(y) = g'(f(x))(f'(x)(y))
+
+* Linear operators are their own derivatives. If :math:`f` linear, then
+
+.. math::
+   f'(x)(y) = f(y)
+
+Implementations in ODL
+~~~~~~~~~~~~~~~~~~~~~~
+
+* The derivative is implemented in ODL  for `Operator`'s via the `Operator.derivative` method.
+* It can be numerically computed using the `NumericalDerivative` operator.
+* Many of the operator arithmetic classes implement the usual rules for the Fréchet derivative, such as the chain rule, distributivity over addition etc.
+
+Gradient
+########
+In the classical setting of functionals :math:`f : \mathbb{R}^n \rightarrow \mathbb{R}`, the gradient is the vector
+
+.. math::
+    \nabla f =
+    \begin{bmatrix}
+        \dfrac{\partial f}{\partial x_1}
+        , \dots,
+        \dfrac{\partial f}{\partial x_n}
+    \end{bmatrix}
+
+This can be generalized to the setting of functionals :math:`f : \mathcal{X} \rightarrow \mathbb{R}` mapping elements in some Hilbert space :math:`\mathcal{X}` to the real numbers by noting that the derivative has a special form by the `Riesz representation theorem
+<https://en.wikipedia.org/wiki/Riesz_representation_theorem>`_, namely that the linear operator :math:`f'(x)` can be represented by a vector :math:`[\nabla f](x) \in \mathcal{X}` such that for all :math:`y \in \mathcal{X}`:
+
+.. math::
+    f'(x)(y) = \langle y, [\nabla f](x) \rangle.
+
+We call the (possibly nonlinear) operator :math:`x \rightarrow [\nabla f](x)` the *Gradient operator* of :math:`f`.
+
+Implementations in ODL
+~~~~~~~~~~~~~~~~~~~~~~
+
+* The gradient is implemented in ODL `Functional`'s via the `Functional.gradient` method.
+* It can be numerically computed using the `NumericalGradient` operator.
```

Comment:
Since it is intended as a listing of the various places where the concept is implemented.